guardian/prosemirror-typerighter
# chore: Use vite

**NB:** depends on #209 – please review that first 🙏 

## What does this change?

Swaps out rollup in favour of [vite](https://vitejs.dev/) to handle our dev and production builds. (Vite uses rollup under the hood for the build process.)

## Why?

Working with this library locally, the current, rollup-based workflow takes a very long time to reflect your changes in a development environment (between 10-20s) Once the build is ready, you must refresh your browser to see changes.

Vite's built to improve that experience:
- the build system uses ESM directly, rebuilding single files on the fly and serving them directly to the browser
- Vite will swap out the smallest portion of code possible when changes are made, statelessly altering your app on the fly to reflect changes to your app:

![hmr](https://user-images.githubusercontent.com/7767575/176470689-69bb2d97-7094-4410-9e40-d019cbf4b8cc.gif)

Vite doesn't provide on-the-fly typechecking out of the box, so we've added it by forking the server process and running `tsc` in watch mode in parallel via `run-p`.

## How to test

**NB:** As our dev-nginx config has had to change to enable Vite to talk to the site over websockets, I've taken the opportunity to change the local domain for consistency with other services. Update your  dev-nginx configuration with `dev-nginx setup-app ./nginx/nginx-mapping.yaml` before running these changes. The app now runs at `https://prosemirror-typerighter.local.dev-gutools.co.uk`. 

- Run the application locally and make a few changes to exercise the HMR and typechecking. Does the application behave as expected?
- Build `prosemirror-typerighter` locally, and use it in a consuming project, e.g. `flexible-content`, linking it via [yalc (or similar).](https://github.com/wclr/yalc). It should continue to work as normal. (Very happy to pair on this!)
- Todo: I'll issue a beta release for this build to verify this over npm, too.
